app/controllers/course/assessment/programming_evaluations_controller.rb

```diff
@@ -79,12 +78,24 @@ def load_programming_evaluation
 
   # Obtains a programming evaluation task accessible by and suitable for the current user.
   #
-  # @return [Course::Assessment::ProgrammingEvaluation|nil] The programming evaluation.
+  # This uses two database queries -- the first to shortlist, the second to lock. Ideally, this
+  # should be one atomic operation, but Postgres does not support row-level locking when used
+  # with DISTINCT.
+  #
+  # @raise [IllegalStateError] If the retrieved record is already assigned to another evaluator
+  #   when locking. The caller should retry finding another evaluation, in that case.
+  # @return [Course::Assessment::ProgrammingEvaluation] The programming evaluation.
   # @return [nil] If no evaluations are found.
   def find_pending_programming_evaluation
-    Course::Assessment::ProgrammingEvaluation.
-      accessible_by(current_ability, :show).
-      with_language(language_param).
-      pending.limit(1).first
+    evaluation = Course::Assessment::ProgrammingEvaluation.
+                 accessible_by(current_ability, :show).
+                 with_language(language_param).
+                 pending.take
+    return nil unless evaluation
+
+    evaluation = evaluation.lock!
```

Comment:
`lock!` uses a `FOR UPDATE` clause, which in Postgres, is a row-level lock:

> 

    FOR UPDATE causes the rows retrieved by the SELECT statement to be locked as though for update. This prevents them from being locked, modified or deleted by other transactions until the current transaction ends. That is, other transactions that attempt UPDATE, DELETE, SELECT FOR UPDATE, SELECT FOR NO KEY UPDATE, SELECT FOR SHARE or SELECT FOR KEY SHARE of these rows will be blocked until the current transaction ends; conversely, SELECT FOR UPDATE will wait for a concurrent transaction that has run any of those commands on the same row, and will then lock and return the updated row (or no row, if the row was deleted). Within a REPEATABLE READ or SERIALIZABLE transaction, however, an error will be thrown if a row to be locked has changed since the transaction started. For further discussion see Section 13.4.
>
> The FOR UPDATE lock mode is also acquired by any DELETE on a row, and also by an UPDATE that modifies the values on certain columns. Currently, the set of columns considered for the UPDATE case are those that have a unique index on them that can be used in a foreign key (so partial indexes and expressional indexes are not considered), but this may change in the future.